src/middlewared/middlewared/plugins/acme_protocol.py

```diff
@@ -0,0 +1,366 @@
+        registration_id = self.middleware.call_sync(
+            'datastore.insert',
+            self._config.datastore,
+            {
+                'uri': register.uri,
+                'tos': register.terms_of_service,
+                'new_account_uri': directory.newAccount,
+                'new_nonce_uri': directory.newNonce,
+                'new_order_uri': directory.newOrder,
+                'revoke_cert_uri': directory.revokeCert,
+                'directory': data['acme_directory_uri']
+            }
+        )
+
+        # Save registration body
+        self.middleware.call_sync(
+            'datastore.insert',
+            'system.acmeregistrationbody',
+            {
+                'contact': register.body.contact[0],
+                'status': register.body.status,
+                'key': key.json_dumps(),
+                'acme': registration_id
+            }
+        )
+
+        return self.middleware.call_sync(f'{self._config.namespace}._get_instance', registration_id)
+
+
+class DNSAuthenticatorService(CRUDService):
+
+    class Config:
+        namespace = 'acme.dns.authenticator'
+        datastore = 'system.acmednsauthenticator'
+
+    def __init__(self, *args, **kwargs):
+        super(DNSAuthenticatorService, self).__init__(*args, **kwargs)
+        self.schemas = DNSAuthenticatorService.initialize_authenticator_schemas()
+
+    @accepts()
+    def authenticator_schemas(self):
+        return [
+            {'schema': [v.to_json_schema() for v in value], 'key': key}
+            for key, value in self.schemas.items()
+        ]
+
+    @staticmethod
+    @private
+    def initialize_authenticator_schemas():
+
+        return {
+            f_n[len('update_txt_record_'):]: [
+                Str(arg, required=True)
+                for arg in list(getattr(DNSAuthenticatorService, f_n).__code__.co_varnames)
+                [4: getattr(DNSAuthenticatorService, f_n).__code__.co_argcount]
+            ]
+            for f_n in [
+                func for func in dir(DNSAuthenticatorService)
+                if callable(getattr(DNSAuthenticatorService, func)) and func.startswith('update_txt_record_')
+            ]
+        }
+
+    @private
+    async def common_validation(self, data, schema_name):
+        verrors = ValidationErrors()
+        if data['authenticator'] not in self.schemas:
+            verrors.add(
+                f'{schema_name}.authenticator',
+                f'System does not support {data["authenticator"]} as an Authenticator'
+            )
+        else:
+            verrors = validate_attributes(self.schemas[data['authenticator']], data)
+
+        if verrors:
+            raise verrors
+
+    @accepts(
+        Dict(
+            'dns_authenticator_create',
+            Str('authenticator', required=True),
+            Str('name', required=True),
+            Dict('attributes', additional_attrs=True, required=True)
+        )
+    )
+    async def do_create(self, data):
```

Comment:
We need to start documenting our APIs as we go.

Can you add a docstring for the public methods?

Check out how its done in pool plugin, please.